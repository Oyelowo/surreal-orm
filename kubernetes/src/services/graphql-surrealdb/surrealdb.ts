import { ServiceDeployment } from '../../shared/deployment.js';
import { AppConfigs, AppEnvVars, TikVDbEnvVars } from '../../types/ownTypes.js';
import { getIngressUrl } from '../../infrastructure/ingress/hosts.js';
import { PlainSecretsManager } from '../../../scripts/utils/plainSecretsManager.js';
import { getEnvVarsForKubeManifests } from '../../shared/environmentVariablesForManifests.js';
import { graphqlSurrealdb } from './app.js';

const env = getEnvVarsForKubeManifests();

const secrets = new PlainSecretsManager('services', 'graphql-surrealdb', 'local').getSecrets();

type SurrealDbEnvVars = AppEnvVars & TikVDbEnvVars<'applications'>;

const surrealDbEnvVars: SurrealDbEnvVars = {
    APP_ENVIRONMENT: env.ENVIRONMENT,
    APP_HOST: '0.0.0.0',
    APP_PORT: '8000',
    APP_EXTERNAL_BASE_URL: getIngressUrl({ environment: env.ENVIRONMENT }),
    TIKV_HOST: 'tikv-pd.applications',
    TIKV_NAME: 'tikv',
    TIKV_PORT: '2379',
    TIKV_SERVICE_NAME: 'tikv',
    TIKV_STORAGE_CLASS: 'linode-block-storage-retain',
};

// Surrealdb is a compute/logic protocol layer using TiKV as persistent layer
// so, we can also deploy it as a kubernetes deployment or statefulset
export const surrealdbSettings: AppConfigs<'surrealdb', 'applications', SurrealDbEnvVars> = {
    kubeConfig: {
        requestMemory: '70Mi',
        requestCpu: '100m',
        limitMemory: '200Mi',
        limitCpu: '100m',
        replicaCount: 2,
        readinessProbePort: 8000,
        host: '0.0.0.0',
        image: `surrealdb/surrealdb:1.0.0-beta.8`,
        command: ['/surreal'],
        // TODO: Change hardcoded credentials once surrealdb is 1.0
        commandArgs: [
            'start',
            '--log',
            'debug',
            '--user',
            secrets.SURREALDB_ROOT_USERNAME,
            '--pass',
            secrets.SURREALDB_ROOT_PASSWORD,
            `tikv://${surrealDbEnvVars.TIKV_HOST}:${surrealDbEnvVars.TIKV_PORT}`,
        ],
    },
    envVars: surrealDbEnvVars,
    metadata: {
        name: 'surrealdb',
        namespace: 'applications',
    },
};

export const surrealDbDeployment = new ServiceDeployment('surrealdb', surrealdbSettings);
surrealDbDeployment.setProvider(graphqlSurrealdb.getProvider());
