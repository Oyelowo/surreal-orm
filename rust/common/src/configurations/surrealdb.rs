use super::utils::get_env_vars_by_prefix;
use serde::Deserialize;
use serde_aux::prelude::deserialize_number_from_string;
use surrealdb::Datastore;

#[derive(Deserialize, Debug, Clone)]
#[serde(rename_all = "lowercase")]
pub struct SurrealdbConfigs {
    pub namespace: String,
    pub database: String,
    pub username: String,
    pub password: String,
    pub host: String,

    #[serde(deserialize_with = "deserialize_number_from_string")]
    pub port: u16,

    #[serde(default = "default_require_ssl")]
    pub require_ssl: Option<bool>,
}

fn default_require_ssl() -> Option<bool> {
    Some(false)
}

impl Default for SurrealdbConfigs {
    fn default() -> Self {
        get_env_vars_by_prefix("SURREALDB_")
    }
}

impl SurrealdbConfigs {
    pub fn get_database(self) -> anyhow::Result<Datastore> {
        // let ds = Datastore::new("tikv://127.0.0.1:8000").await?;
        todo!()
    }
}
