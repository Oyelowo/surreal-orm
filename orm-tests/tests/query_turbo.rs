/*
 * Author: Oyelowo Oyedayo
 * Email: oyelowo.oss@gmail.com
 * Copyright (c) 2023 Oyelowo Oyedayo
 * Licensed under the MIT license
 */

use pretty_assertions::assert_eq;
use surreal_models::{account, Account, Balance};
use surreal_orm::{
    statements::{create, create_only, select, update},
    *,
};
use surrealdb::{engine::local::Mem, Surreal};

#[tokio::test]
async fn test_simple_standalone_for_loop() -> SurrealOrmResult<()> {
    let account::Schema { balance, .. } = Account::schema();
    let account_table = Account::table_name();

    let query = query_turbo! {
        for name in vec!["Oyelowo", "Oyedayo"] {
            let new_bal = 5;
            select(All).from(account_table).where_(balance.eq(new_bal));
        };
    };

    insta::assert_display_snapshot!(query.to_raw().build());
    insta::assert_display_snapshot!(query.fine_tune_params());

    Ok(())
}

#[tokio::test]
async fn test_multiple_for_loops() -> SurrealOrmResult<()> {
    let account::Schema { balance, .. } = Account::schema();
    let account_table = &Account::table_name();

    let names_closure = || vec!["Oyelowo", "Oyedayo"];

    let query = query_turbo! {
        for name in vec!["Oyelowo", "Oyedayo"] {
            let new_bal = 5;
            select(All).from(account_table).where_(balance.eq(new_bal));

            let names = vec!["Oyelowo", "Oyedayo"];
            for name in names {
                let amount_to_use = 999;
                create_only::<Balance>().set(object!(Balance {
                    id: Balance::create_id("balance1".to_string()),
                    amount: amount_to_use,
                }));
            };
        };

        for name in  names_closure() {
            let new_bal = 5;
            select(All).from(account_table).where_(balance.eq(new_bal));
        };
    };

    insta::assert_display_snapshot!(query.to_raw().build());
    insta::assert_display_snapshot!(query.fine_tune_params());

    Ok(())
}

// with simple standalone if else statement
#[tokio::test]
async fn test_simple_standalone_if_else() -> SurrealOrmResult<()> {
    let account::Schema { balance, .. } = Account::schema();
    let account_table = &Account::table_name();

    let second_cond_from_outside = balance.lte(54);
    // let xx = {
    // };

    let query = query_turbo! {
        let first_cond = balance.gte(100);
        if first_cond {
            let first_name = "Oyelowo";
            let score = 100;
            select(All).from(account_table).where_(balance.eq(5));
        } else if second_cond_from_outside {
            let first_name = "Oyelowo";
            let score = 100;
            select(All).from(account_table).where_(balance.eq(5));
        } else if (balance.gte(100)) {
           let first_name = "Oyelowo";
            let score = 100;
            select(All).from(account_table).where_(balance.eq(5));
        } else {
            let first_name = "Oyelowo";
            let score = 100;
            select(All).from(account_table).where_(balance.eq(5));
        };
    };

    insta::assert_display_snapshot!(query.to_raw().build());
    insta::assert_display_snapshot!(query.fine_tune_params());

    Ok(())
}

#[tokio::test]
async fn test_multiple_if_else() -> SurrealOrmResult<()> {
    let account::Schema { balance, .. } = Account::schema();

    let query = query_turbo! {
        let within_turbo_cond = balance.equal(33);
        let cond_username = if  within_turbo_cond {
            let first_name = "Oyelowo";
            if balance.equal(33) {
                let username = "oye";
                if balance.equal(92) {
                    let username = "Oyedayo";
                    if within_turbo_cond {
                        let username = "codebreather";
                    };

                    for name in vec!["Oyelowo", "Oyedayo"] {
                        let first = "Oyelowo";
                        select(All).from(Account::table_name()).where_(balance.eq(5));
                    };

                };
            } else {
                let score = 100;
                select(All).from(Account::table_name()).where_(balance.eq(5));
            };
        } else if balance.less_than(100) {
            let first_name = "Oyelowo";
            let score = 100;
            select(All).from(Account::table_name()).where_(balance.eq(5));

        } else if (balance.gte(100)) {
           let first_name = "Oyelowo";
            let score = 100;
            return select(All).from(Account::table_name()).where_(balance.eq(5));
        } else {
            let first_name = "Oyelowo";
            let score = 100;
            select(All).from(Account::table_name()).where_(balance.eq(5));
        };
        let score = 100;

        select(All).from(Account::table_name()).where_(balance.eq(5));

        if (balance.greater_than(100)) {
            let first_name = "Oyelowo";
        };

        return cond_username;
    };

    insta::assert_display_snapshot!(query.to_raw().build());
    insta::assert_display_snapshot!(query.fine_tune_params());

    Ok(())
}

#[tokio::test]
async fn test_transaction_with_block_macro() -> SurrealOrmResult<()> {
    let db = Surreal::new::<Mem>(()).await.unwrap();
    db.use_ns("test").use_db("test").await.unwrap();

    let id1 = &Account::create_id("one".to_string());
    let id2 = &Account::create_id("two".to_string());
    let amount_to_transfer = 300.00;

    let acc = Account::schema();
    let account::Schema { balance, .. } = Account::schema();

    let query_chain = query_turbo! {
        begin transaction;
         let balance1 = create_only().content(Balance {
                id: Balance::create_id("balance1".to_string()),
                amount: amount_to_transfer,
            });

         create_only().content(Balance {
                id: Balance::create_id("balance2".to_string()),
                amount: amount_to_transfer,
            });

        if balance.greater_than(100) {
            let first_name = "Oyelowo";
            let score = 100;
            select(All).from(Account::table_name()).where_(acc.balance.eq(5));
        } else if balance.less_than(100) {
            let first_name = "Oyelowo";
            let score = 100;
            select(All).from(Account::table_name()).where_(acc.balance.eq(5));
        } else if balance.gte(100) {
            let first_name = "Oyelowo";
            let score = 100;
            select(All).from(Account::table_name()).where_(acc.balance.eq(5));
        } else {
            let first_name = "Oyelowo";
            let score = 100;
            select(All).from(Account::table_name()).where_(acc.balance.eq(5));
        };

        // You can also use parenthesis for the iteration if you want
        // for (name in vec!["Oyelowo", "Oyedayo"]) {
        // let names = vec!["Oyelowo", "Oyedayo"];
        // for name in  names {
        for name in vec!["Oyelowo", "Oyedayo"] {
            let first = "Oyelowo";
            select(All).from(Account::table_name()).where_(acc.balance.eq(5));

            let good_stmt = select(All).from(Account::table_name()).where_(acc.balance.eq(64));

            if balance.gt(50) {
                let first_name = "Oyelowo";
            };

            select(All).from(Account::table_name()).where_(acc.balance.eq(34));

            let numbers = vec![23, 98];

            for age in numbers {
              let score = 100;
              let first_stmt = select(All).from(Account::table_name()).where_(acc.balance.eq(5));

              let second_stmt = select(All).from(Account::table_name()).where_(acc.balance.eq(25));
              select(All).from(Account::table_name()).where_(acc.balance.eq(923));

            };
        };

         let  balance3 = create().content(Balance {
                id: Balance::create_id("balance3".into()),
                amount: amount_to_transfer,
            });

        let accounts = select(All)
            .from(id1..=id2);


        // You can reference the balance object by using the $balance variable and pass the amount
        // as a parameter to the decrement_by function. i.e $balance.amount
        let updated1 = update::<Account>(id1).set(acc.balance.increment_by(balance1.with_path::<Balance>(E).amount));
        update::<Account>(id1).set(acc.balance.increment_by(balance1.with_path::<Balance>(E).amount));
        update::<Account>(id1).set(acc.balance.increment_by(45.3));

        // You can also pass the amount directly to the decrement_by function. i.e 300.00
        let update2 = update::<Account>(id2).set(acc.balance.decrement_by(amount_to_transfer));
        // FIX: Scenario where the same variable is used twice, only the latest let statement value
        // is duplicated because we are using the bound variable name in the chaining rather than
        // the value itself. So, either probably consider suffixing the variable name with a random
        // id or using the values directly in the binding.
        // Or probably doing the chaining immediately after declaration and aggregating that after
        // the entire chained queries.
        let update3 = update::<Account>(id2).set(acc.balance.decrement_by(50));

        commit transaction;
    };

    insta::assert_display_snapshot!(query_chain.to_raw().build());
    insta::assert_display_snapshot!(query_chain.fine_tune_params());

    let result = query_chain.run(db.clone()).await?;

    let accounts = select(All)
        .from(id1..=id2)
        .return_many::<Account>(db.clone())
        .await?;

    assert_eq!(accounts.len(), 2);
    assert_eq!(accounts[0].balance, 645.3);
    assert_eq!(accounts[1].balance, -350.0);
    assert_eq!(accounts[0].id.to_string(), "account:one");
    assert_eq!(accounts[1].id.to_string(), "account:two");
    Ok(())
}
