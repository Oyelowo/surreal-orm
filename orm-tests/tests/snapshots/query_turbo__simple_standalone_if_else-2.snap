---
source: orm-tests/tests/query_turbo.rs
expression: query.fine_tune_params()
---
LET $first_cond = balance >= $_param_00000001;

IF $first_cond 
	{ LET $first_name = $_param_00000002;

LET $score = $_param_00000003;

SELECT * FROM account WHERE balance = $_param_00000004; }
ELSE IF balance <= $_param_00000005 
	{ LET $first_name = $_param_00000006;

LET $score = $_param_00000007;

SELECT * FROM account WHERE balance = $_param_00000008; }
ELSE IF balance >= $_param_00000009 
	{ LET $first_name = $_param_00000010;

LET $score = $_param_00000011;

SELECT * FROM account WHERE balance = $_param_00000012; }
ELSE
	{ LET $first_name = $_param_00000013;

LET $score = $_param_00000014;

SELECT * FROM account WHERE balance = $_param_00000015; };
